matrix/rules.py

```diff
@@ -364,12 +369,11 @@ def allow_event(e):
             try:
                 await self.add_model(context)
                 await self.run_once(context, test)
-            except model.TestFailure as e:
-                self.handle_shutdown(None)
-                if self.fail_fast and e.task.gating is True:
-                    log.info("Fail Fast on Test failure: %s" % test.name)
-                    break
             finally:
```

Comment:
The TestFailure gets caught and handled elsewhere (in the asyncio lib), so this code never gets executed. Chopping it out so that it won't confuse future devs as much as it confused me :-)